source/docs/software/basic-programming/java-gc.rst

```diff
@@ -32,14 +32,50 @@ If you anticipate your application creating a large number of short-lived object
+   :lineno-start: 15
+   :emphasize-lines: 15-16
+
+Add to the code block so that it contains two ``jvmArgs`` commands, as shown below:
+
+.. code-block:: groovy
+
+   frcJava(getArtifactTypeClass('FRCJavaArtifact')) {
+       // If you have other configuration here, you do not need to remove it.
+       // Enable automatic heap dumps on OutOfMemoryError
+       jvmArgs.add("-XX:+HeapDumpOnOutOfMemoryError")
+       jvmArgs.add("-XX:HeapDumpPath=/home/lvuser/frc-usercode.hprof")
+   }
+
+This will cause the JVM to write heap dumps to a file named ``frc-usercode.hprof`` in ``/home/lvuser`` on the roboRIO when the code runs out of memory. :doc:`Copy the file from the roboRIO to a computer using FTP/SFTP </docs/software/roborio-info/roborio-ftp>` and delete it from the roboRIO. You can then open this file in a memory profiler such as :ref:`VisualVM <docs/software/advanced-gradlerio/profiling-with-visualvm:Analyzing a Heap Dump>`.
+
+.. warning:: There are several things to consider when using these options:
+
+   - If the file specified by the path in the ``-XX:HeapDumpPath`` option already exists, the JVM will **not** overwrite it if the program runs out of memory again, nor will it write the heap dump to a file with a different name.
+   - Heap dumps intrinsically consume space equal to the amount of memory the program had allocated when it ran out of memory. This can lead to (relatively) large files consuming most of the roboRIO's available persistent storage, which may interfere with the ability to deploy code, among other things. **Always** delete heap dumps from the roboRIO once you have copied them to your computer.
+
+     - Alternatively, you can insert a USB flash drive into to roboRIO and alter the ``-XX:HeapDumpPath`` argument to point to a location on the flash drive, which can typically be found at ``/u``. If you do this, **the flash drive MUST remain connected to the roboRIO while your code is running.** Once you have collected the heap dump, redeploy your code with the heap dump options removed.
+
+   Considering these limitations, it is not recommended to use these options during competitive play.
+
+System Memory Tuning
+--------------------
 
 If the JVM cannot allocate memory, the program will be terminated. As an embedded system with only a small amount of memory available (256 MB on the roboRIO 1, 512 MB on the roboRIO 2), the roboRIO is particularly susceptible to running out of memory.
 
 .. admonition :: No amount of system tuning can fix out of memory errors caused by out-of-control allocations.
 
-    If you are running out of memory, always investigate allocations with :doc:`VisualVM </docs/software/advanced-gradlerio/profiling-with-visualvm>` first.
+    If you are running out of memory, always investigate allocations with :ref:`heap dumps <docs/software/basic-programming/java-gc:diagnosing out of memory errors with heap dumps>` and/or :doc:`VisualVM </docs/software/advanced-gradlerio/profiling-with-visualvm>` first.
```

Comment:
Since this links to something right above this, and generating a heap dump without analyzing it doesn't seem to add much value, I don't think the link needs to be added hear. But I think it would be value added for the Visual VM page.

Reply:
Should be done as of the last commit.